File: Day88/Part5/multi_agent_merge.cpp
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <queue>
#include <algorithm>
#include <tuple>

using namespace std;

// Problem: Multi-Agent Dungeon With Hostile Merge
// K agents. Move simultaneously.
// If two agents meet (same node), they "merge".
// Merging typically is bad (loss of agent / worse position).
// Let's assume merge -> agents become 1 agent (loss of redundancy).
// Goal: Someone reaches Target.
// State: Sorted tuples of agent positions.
// K is small.
// Grid size N x M.
// BFS on State Space.

int main() {
    int N, M, K;
    if (!(cin >> N >> M >> K)) return 0;

    int inputSize = N * M;
    vector<int> starts(K);
    int Target = -1;

    for (int i = 0; i < K; ++i) {
        int r, c;
        cin >> r >> c;
        starts[i] = r * M + c;
    }

    int tr, tc;
    cin >> tr >> tc;
    Target = tr * M + tc;

    vector<string> grid(N);
    for (int i = 0; i < N; ++i) cin >> grid[i];

    // BFS
    // State is sorted vector of positions.
    // Use set for visited.
    queue<vector<int>> q;
    
    sort(starts.begin(), starts.end());
    q.push(starts);
    
    set<vector<int>> visited;
    visited.insert(starts);

    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};

    while (!q.empty()) {
        vector<int> current_agents = q.front();
        q.pop();

        // Check winner
        for (int pos : current_agents) {
            if (pos == Target) {
                cout << "Possible" << endl;
                return 0;
            }
        }

        // Generate next states
        // All agents move simultaneously.
        // If K agents, and 4 moves each -> 4^K next states.
        // If K=3, 64 next states. Manageable.
        // Need to iterate combinations.
        
        int num_agents = current_agents.size();
        // Helper to recusively generate moves
        vector<vector<int>> next_states_raw;
        // ... (complex recursion needed)
        
        // Let's implement iterative approach for small K.
        // Since input might use K generic, but usually K <= 3 in these problems.
        // I'll assume K is small.
        
        // Actually, "Hostile Merge" usually implies we want to avoid it if possible?
        // Or is it forced?
        // "Jika dua agent bertemu -> hostile merge".
        // This happens AFTER move.
        // So we generate moves -> Check duplicates -> Merge duplicates.
        
        // Function to expand
        vector<vector<int>> current_level_moves;
        current_level_moves.push_back({}); // Start with empty move set
        
        for (int i = 0; i < num_agents; ++i) {
            int u = current_agents[i];
            int r = u / M;
            int c = u % M;
            
            vector<vector<int>> next_level_moves;
            
            for (const auto& partial : current_level_moves) {
                for (int dir = 0; dir < 4; ++dir) {
                    int nr = r + dr[dir];
                    int nc = c + dc[dir];
                    
                    if (nr >= 0 && nr < N && nc >= 0 && nc < M && grid[nr][nc] != '#') {
                        vector<int> next_partial = partial;
                        next_partial.push_back(nr * M + nc);
                        next_level_moves.push_back(next_partial);
                    }
                }
            }
            current_level_moves = next_level_moves;
        }
        
        for (auto& next_pos_list : current_level_moves) {
            // Process merges (sort and unique)
            if (next_pos_list.size() != num_agents) continue; // Should allow stuck agents? Problem typically assumes valid moves.
            // If an agent has no valid move, does the whole team die? Or just that agent?
            // "Agent bergerak simultan". If one stuck, typically invalid state for team.
            
            sort(next_pos_list.begin(), next_pos_list.end());
            next_pos_list.erase(unique(next_pos_list.begin(), next_pos_list.end()), next_pos_list.end());
            
            if (visited.find(next_pos_list) == visited.end()) {
                visited.insert(next_pos_list);
                q.push(next_pos_list);
            }
        }
    }

    cout << "Impossible" << endl;

    return 0;
}

File: Day88/Part5/multi_agent_merge.py
import sys
import collections
import itertools

def solve():
    try:
        line1 = sys.stdin.readline().split()
        if not line1: return
        N, M, K = map(int, line1)
        
        starts = []
        for _ in range(K):
            starts.append(tuple(map(int, sys.stdin.readline().split()))) # (r, c)
            
        line_t = sys.stdin.readline().split()
        if not line_t: return
        Target = tuple(map(int, line_t))
        
        grid = []
        for _ in range(N):
            grid.append(sys.stdin.readline().strip())
    except ValueError:
        return

    # BFS
    # State: sorted tuple of (r,c)
    initial_state = sorted(starts)
    initial_tuple = tuple(initial_state)
    
    queue = collections.deque([initial_tuple])
    visited = {initial_tuple}
    
    dr = [0, 0, 1, -1]
    dc = [1, -1, 0, 0]
    
    while queue:
        current_agents = queue.popleft()
        
        if any(pos == Target for pos in current_agents):
            print("Possible")
            return
            
        # Generate moves
        agent_options = []
        possible = True
        
        for r, c in current_agents:
            opts = []
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if 0 <= nr < N and 0 <= nc < M and grid[nr][nc] != '#':
                    opts.append((nr, nc))
            
            if not opts:
                possible = False
                break
            agent_options.append(opts)
            
        if not possible:
            continue
            
        for next_config in itertools.product(*agent_options):
            # Sort and merge duplicates
            unique_config = sorted(list(set(next_config)))
            next_state = tuple(unique_config)
            
            if next_state not in visited:
                visited.add(next_state)
                queue.append(next_state)
                
    print("Impossible")

if __name__ == "__main__":
    solve()
